### PR TITLE
GPT training model with nn.Graph

### PR DESCRIPTION
**Pull request by leaves-zwx**
需要导出的 op functor (需要 @doombeaker 帮忙计划下，有疑问可找我讨论和 review)：

- [x] sparse_softmax_cross_entropy_with_logits（裴廷宽）
- [x] fused_scale_tril（张文骁）
- [ ] fused_scale_tril_softmax_dropout（李春游，已经完成初稿，待review）
- [x] fused_bias_add_gelu(郑泽康）
- [x] fused_bias_add_dropout（郑泽康）
- [ ] fused_self_attention_query_mul_key_and_value（姚迟）

> 优先级从上至下

需要支持的机制：
- pipeline stage @strint @chengtbf 
- 亚线性优化（checkpoint_activations）  @tingkuanpei 
- amp_white_identity 的新接口 @luqiang-guo 

旧的 function_config 中的选项需要在 nn.Graph 中暴露

```python
    func_cfg = flow.function_config()
    if args.fp16:
        func_cfg.enable_auto_mixed_precision(True)
    func_cfg.prune_parallel_cast_ops(True)
    func_cfg.enable_fuse_add_to_output(True)
    func_cfg.enable_fuse_model_update_ops(True)
    func_cfg.enable_fuse_cast_scale(True)
    # turn on this flag when match ZeRO & DeepSpeed
    func_cfg.enable_non_distributed_optimizer(False)
```

### CODE REVIEW

**Comment by doombeaker** [comment]
这个 `sparse_softmax_cross_entropy_with_logits` 算子： https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/compatible/single_client/ops/nn_ops.py#L2399

我看它内部其实就是根据 sbp 判断走了不同的分支：

```python
if logits.distribute is oneflow._oneflow_internal.distribute.split(
        len(logits.shape) - 1
    ):
        走这个op：out = "sparse_softmax_cross_entropy_ms"
    else:
        走这个op：out = "sparse_softmax_cross_entropy"
    return out
```

而对于后者，其实我们已经有对用的 Module `nn.CrossEntropyLoss` 了，它的内部调用的是 `sparse_softmax_cross_entropy`。

https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/nn/modules/loss.py#L198
->
https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/core/functional/impl/nn_functor.cpp#L362

现有的 `nn.CrossEntropyLoss` 能否满足要求呢？还是说还是需要按照原有 lazy，根据sbp不同走不同分支，才符合要求。

**Comment by leaves-zwx** [comment]
> 现有的 nn.CrossEntropyLoss 能否满足要求呢？还是说还是需要按照原有 lazy，根据sbp不同走不同分支，才符合要求。

我觉得暴露 oneflow.F. sparse_softmax_cross_entropy_ms 就可以了

**Comment by leaves-zwx** [comment]
配置对齐后跑单卡 117M 模型的训练，300轮的 loss 曲线对比

image

graph 版本下降教慢，原因还在找。

吞吐方面在类脑机器上测试 batch_size == 8 的时候 benchmark 中的吞吐大概是 38 samples/s，新版 graph 的吞吐大概是 31 samples/s，估计是有些优化没开启（fusion 类的）

**Comment by leaves-zwx** [comment]
image

单卡的 loss 可以对齐了，问题是出在 https://github.com/Oneflow-Inc/models/pull/109/commits/c4829458f230377440fc6a3ce9d9f1cc8f1dac75#diff-e92db12435665cf49b5838dde444782e2f19dba9c3b9e669414f55526e6f2675R53 这里，为 optimizer 添加 param_groups 的时候没有加 embedding 的参数，导致 embedding 的 wpe 和 wte 没更新。

**Comment by leaves-zwx** [comment]
```
scale_tril_softmax_dropout_fusion ............... False
multihead_attention_fusion ...................... False
```
关闭了这两个 fusion 后（graph 下还没支持）

300 轮的吞吐对比（10轮打印一次）

image

大致一样，但 graph 有一些抖动，推测可能是 print(tensor) 引起的同步问题

**Comment by strint** [comment]
```
   func_cfg = flow.function_config()
    if args.fp16:
        func_cfg.enable_auto_mixed_precision(True)     # enable_amp
    func_cfg.prune_parallel_cast_ops(True)  #   在这个pr加 https://github.com/Oneflow-Inc/oneflow/pull/6415
    func_cfg.enable_fuse_add_to_output(True)  #  allow_fuse_add_to_output
    func_cfg.enable_fuse_model_update_ops(True)  # allow_fuse_model_update_ops
    func_cfg.enable_fuse_cast_scale(True)  # allow_fuse_cast_scale
    # turn on this flag when match ZeRO & DeepSpeed
    # 这里是false的话，应该不用设置。在这个pr加 https://github.com/Oneflow-Inc/oneflow/pull/6415 需要打开时，用 graph.config.set_zero_redundancy_optimizer_mode("distributed_split")
    func_cfg.enable_non_distributed_optimizer(False)   
```

**Comment by ouyangyu** [comment]
<p align="center">表 1-1 Lazy模式下，GPT测试案例 2D_PP_2x4x4_512_2304x24 </p>

| 分支                                                         | lantency      | memory      | throuthput(sample/sec) | 日志                                                         |
| ------------------------------------------------------------ | ------------- | ----------- | ---------------------- | ------------------------------------------------------------ |
| [v0.3.0](https://github.com/Oneflow-Inc/oneflow/releases/tag/v0.3.0) | 13117.18 (ms) | 11958 (MiB) | 39.03                  | [log](https://oneflow-public.oss-cn-beijing.aliyuncs.com/GPT/perf_test_logs/pretrain_gpt_4n8d_dp2_mp4_pp4_mbz8_gbz512_s2048_l24_h2304_nh24.log) |
| v0.5rc2的修复分支[dev_rm_check_for_0d_tensor](https://github.com/Oneflow-Inc/oneflow/tree/dev_rm_check_for_0d_tensor) | 13326.51(ms)  | 12630 (MiB) | 38.41                  | [log](https://oneflow-test.oss-cn-beijing.aliyuncs.com/v0.5.0_release_model_log/GPT/pretrain_gpt_4n8d_dp2_mp4_pp4_mbz8_gbz512_s2048_l24_h2304_nh24.log) |

**Comment by leaves-zwx** [comment]
目前 4 机情况下因为还有 bug 没有跑通。今天之后没有类脑 4 机 32 卡的环境可以使用，且暂时没有可替代的测试环境了，将寻求其他的测试 case 在合适的测试机器上测试。

最基本的完整测试 gpt 的并行策略的测试环境至少需要 3 机 12 卡。（pipeline 3机 + 2d 4卡 ）

**Review by Ldpe2G — APPROVED** [review]
